core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -3955,11 +3960,473 @@ class KafkaApis(val requestChannel: RequestChannel,
+        if (bandwidthThrottleTimeMs > requestThrottleTimeMs) {
+          requestHelper.throttle(quotas.fetch, request, bandwidthThrottleTimeMs)
+        } else {
+          requestHelper.throttle(quotas.request, request, requestThrottleTimeMs)
+        }
+        // If throttling is required, return an empty response.
+        unconvertedShareFetchResponse = shareFetchContext.throttleResponse(maxThrottleTimeMs)
+      } else {
+        // Get the actual response. This will update the fetch context.
+        unconvertedShareFetchResponse = shareFetchContext.updateAndGenerateResponseData(groupId, Uuid.fromString(memberId), partitions)
+        val responsePartitionsSize = unconvertedShareFetchResponse.data.responses.stream().mapToInt(_.partitions.size()).sum()
+        trace(s"Sending Share Fetch response with partitions size=$responsePartitionsSize")
+      }
+      createResponse(maxThrottleTimeMs)
+    }
+
+    if (interestingWithMaxBytes.isEmpty) {
+      processResponseCallback(Map.empty)
+    } else {
+      // for share fetch from consumer, cap fetchMaxBytes to the maximum bytes that could be fetched without being
+      // throttled given no bytes were recorded in the recent quota window. Trying to fetch more bytes would result
+      // in a guaranteed throttling potentially blocking consumer progress.
+      val maxQuotaWindowBytes = quotas.fetch.getMaxValueInQuotaWindow(request.session, clientId).toInt
+
+      val fetchMaxBytes = Math.min(Math.min(shareFetchRequest.maxBytes, config.fetchMaxBytes), maxQuotaWindowBytes)
+      val fetchMinBytes = Math.min(shareFetchRequest.minBytes, fetchMaxBytes)
+
+      val clientMetadata: Optional[ClientMetadata] =
+        Optional.of(new DefaultClientMetadata(
+          CommonClientConfigs.DEFAULT_CLIENT_RACK,
+          clientId,
+          request.context.clientAddress,
+          request.context.principal,
+          request.context.listenerName.value))
+
+      // TODO : Change the dummy values to actual values after the required implementations are completed
+      val params = new FetchParams(
+        versionId,
+        FetchRequest.FUTURE_LOCAL_REPLICA_ID,
+        -1,
+        shareFetchRequest.maxWait,
+        fetchMinBytes,
+        fetchMaxBytes,
+        FetchIsolation.HIGH_WATERMARK,
+        clientMetadata
+      )
+
+      val interestingTopicPartitions: util.List[TopicIdPartition] = new util.ArrayList[TopicIdPartition]()
+      val iterator = interestingWithMaxBytes.entrySet().iterator()
+      while (iterator.hasNext) {
+        val entry = iterator.next()
+        interestingTopicPartitions.add(entry.getKey)
+      }
+
+      // call the share partition manager to fetch messages from the local replica.
+      try{
+        val responsePartitionData = sharePartitionManager.fetchMessages(
+          groupId,
+          shareFetchRequest.data.memberId,
+          params,
+          interestingTopicPartitions,
+          interestingWithMaxBytes
+        ).get()
```

Comment:
Hmm the whole processing is `synchronous`, why can't we work with callbacks i.e. when complete?

Reply:
I can find this comment is yet not addressed. I think we discussed that the code shall be asynchronous. In case we are finding it difficult in this PR then please log a jira and work on it post merge of this PR.

Reply:
Created a JIRA - https://issues.apache.org/jira/browse/KAFKA-17126